nni/algorithms/compression/tensorflow/pruning/one_shot.py

```diff
@@ -57,15 +62,54 @@ def calc_masks(self, sparsity, wrapper, wrapper_idx=None):
 
             w_abs = tf.math.abs(weight)
             k = tf.size(weight) - num_prune
-            topk = tf.math.top_k(tf.reshape(w_abs, [-1]), k)[0]
+            topk = tf.math.top_k(tf.reshape(w_abs, [-1]), k).values
             if tf.size(topk) == 0:
                 mask = tf.zeros_like(weight)
             else:
                 mask = tf.math.greater_equal(w_abs, topk[-1])
             masks[weight_variable.name] = tf.cast(mask, weight.dtype)
         return masks
 
+class SlimPrunerMasker(WeightMasker):
+    def __init__(self, model, pruner, **kwargs):
+        super().__init__(model, pruner)
+        weight_list = []
+        for wrapper in pruner.wrappers:
+            weights = [w for w in wrapper.layer.weights if '/gamma:' in w.name]
+            assert len(weights) == 1, f'Bad weights: {[w.name for w in wrapper.layer.weights]}'
+            weight_list.append(tf.math.abs(weights[0].read_value()))
+        all_bn_weights = tf.concat(weight_list, 0)
+        k = int(all_bn_weights.shape[0] * (1 - pruner.wrappers[0].config['sparsity']))
```

Comment:
This place is a little different from pytorch implementation and may lead to different thresholds with the same weight and sparsity.

https://github.com/microsoft/nni/blob/761732ab5f8e2606d6b5a5c6e0bf025d3a085b60/nni/algorithms/compression/pytorch/pruning/structured_pruning.py#L684

Reply:
The topk API is different between PyTorch and TensorFlow. If we don't use `1 - ` the logic will be much more complicated and hard to read.
Maybe we can alter the rounding behavior to make them more similar, but I guess the original implementation does not care it at all...

Reply:
Yes, maybe alter the rounding behavior is better. Or each layer may gap at most two channels be pruned between pytorch and tensorflow, then I'm afraid the final effect may have a bigger gap.

Reply:
Maybe we can also use `-tf.nn.top_k(-weights,k,sorted=False).values.numpy()` to get the minimum k values.